### PR TITLE
explicitly set 'include' to numpy_incls

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -460,6 +460,13 @@ def pxd(name):
     return os.path.abspath(pjoin('pandas', name + '.pxd'))
 
 
+if _have_setuptools:
```
Comment (jreback) [inline setup.py]: this whole setuptools optional biz is a mess. we should just require it (and we do in our recipes, so this is all old code). note nothing to do with this PR.

**Comment by jreback** [comment]
this is ok..

File: setup.py [hunk 2]
```diff
@@ -570,14 +580,16 @@ def pxd(name):
     '_libs.tslibs.timezones': {
         'pyxfile': '_libs/tslibs/timezones'},
     '_libs.testing': {
-        'pyxfile': '_libs/testing'},
+        'pyxfile': '_libs/testing',
+        'include': []},
```
Comment (jreback) [inline setup.py]: what is the difference betwee not passing include and an empty list? (from a compilation perspective)?

Comment (gfyoung) [reply setup.py]: I think this can be removed IINM.

Comment (jbrockmendel) [reply setup.py]: Without the 'include' key, it gets set to a default of `common_include`.  This has two unwanted side effects.  First: the search path for numpy now includes src/numpy.pxd which we should move away from.  Second: making it explicit avoids a problem a year from now when we see that `common_include` is added to `_libs.testing` but no one remembers that it isn't actually needed.

**Comment by jreback** [comment]
rebase

**Comment by jbrockmendel** [comment]
How did this happen in master:
```
    '_libs.tslibs.parsing': {
        'pyxfile': '_libs/tslibs/parsing',
        'pxdfiles': ['_libs/src/util',
                     '_libs/src/khash']},
```
?

tslibs.parsing doesnt use either of those.

**Comment by jbrockmendel** [comment]
```
    '_libs.tslibs.timedeltas': {
        'pyxfile': '_libs/tslibs/timedeltas',
        'pxdfiles': ['_libs/src/util'],
        'depends': np_datetime_headers,
        'sources': np_datetime_sources},
```
nope on the depends and sources.

**Comment by jbrockmendel** [comment]
```
    '_libs.tslibs.strptime': {
        'pyxfile': '_libs/tslibs/strptime',
        'pxdfiles': ['_libs/src/util',
                     '_libs/tslibs/nattype'],
        'depends': tseries_depends,
        'sources': np_datetime_sources},
```
nope on depends and sources, missing np_datetime

**Comment by jbrockmendel** [comment]
```
    '_libs.index': {
        'pyxfile': '_libs/index',
        'pxdfiles': ['_libs/src/util', '_libs/hashtable'],
        'depends': _pxi_dep['index'],
        'sources': np_datetime_sources},
```

Nope on the sources

**Comment by jreback** [comment]
np_datetime is now part of tseries_depends

File: setup.py [hunk 3]
```diff
@@ -576,31 +585,29 @@ def pxd(name):
                      '_libs/tslibs/conversion']},
     '_libs.tslibs.parsing': {
         'pyxfile': '_libs/tslibs/parsing',
-        'pxdfiles': ['_libs/src/util',
```
Comment (jreback) [inline setup.py]: pls don’t change this;  if u look in the source these depend on these

File: setup.py [hunk 4]
```diff
@@ -576,31 +585,29 @@ def pxd(name):
...
-                     '_libs/src/khash']},
+        'include': numpy_incls},
     '_libs.tslibs.strptime': {
         'pyxfile': '_libs/tslibs/strptime',
         'pxdfiles': ['_libs/src/util',
-                     '_libs/tslibs/nattype'],
-        'depends': tseries_depends,
```
Comment (jreback) [inline setup.py]: same

File: setup.py [hunk 5]
```diff
@@ -576,31 +585,29 @@ def pxd(name):
...
-        'sources': np_datetime_sources},
+                     '_libs/tslibs/nattype',
+                     '_libs/tslibs/np_datetime']},
     '_libs.tslibs.timedeltas': {
         'pyxfile': '_libs/tslibs/timedeltas',
-        'pxdfiles': ['_libs/src/util'],
-        'depends': np_datetime_headers,
```
Comment (jreback) [inline setup.py]: same

**Review by jreback — CHANGES_REQUESTED** [review]

File: setup.py (hunk 3 above)
Comment (jbrockmendel) [reply setup.py]: Are we looking at different source files?  tslibs.parsing doesn't cimport from util or khash.

File: setup.py (hunk 4 above)
Comment (jbrockmendel) [reply setup.py]: same

File: setup.py (hunk 5 above)
Comment (jbrockmendel) [reply setup.py]: same.

**Comment by jbrockmendel** [comment]
It's almost as if dumping a bunch of unnecessary stuff into setup.py made it easier for people in new and exciting ways.

File: setup.py (hunk 4 above)
Comment (jreback) [reply setup.py]: tseries_depends already incldes np_datetime, so no need to add this explicty.

File: setup.py (hunk 3 above)
Comment (jreback) [reply setup.py]: you are right here

**Review by jreback — CHANGES_REQUESTED** [review]
you need to rebase. some of your comments are on old sources.

**Comment by jreback** [comment]
this looks ok, but why is this necessary? what is wrong with the current?

**Comment by jbrockmendel** [comment]
> this looks ok, but why is this necessary? what is wrong with the current?

Dependencies are over-specified.  For extensions that just have include=numpy_incls, it is explicitly saying "this is not tightly coupled"

Also I really don't like having both numpy_incl and src/numpy.pxd on the search path.

**Comment by jreback** [comment]
> Dependencies are over-specified. For extensions that just have include=numpy_incls, it is explicitly saying "this is not tightly coupled"

can you show how you determine this?

**Comment by jbrockmendel** [comment]
> can you show how you determine this?

I look through the module for cimports.  If the only non-cython/cpython cimports it uses are numpy, then it has no pandas c-deps and the build only needs `np.get_include()`.

File: setup.py (hunk 4 above)
Comment (jreback) [reply setup.py]: i think this one needs to change back

**Review by jreback — CHANGES_REQUESTED** [review]

File: setup.py (hunk 4 above)
Comment (jbrockmendel) [reply setup.py]: From the diff it isn't clear which file you're referring to.  Best guess is tslibs.strptime?

File: setup.py [hunk 6]
```diff
@@ -578,14 +587,12 @@ def pxd(name):
                      '_libs/tslibs/frequencies']},
     '_libs.tslibs.parsing': {
         'pyxfile': '_libs/tslibs/parsing',
-        'pxdfiles': ['_libs/src/util',
```
Comment (jreback) [inline setup.py]: side note, should parsing depend on nattype? (then for example nat strings are in 1 place).

File: setup.py (hunk 4 above)
Comment (jreback) [reply setup.py]: for tslibs.strptime, still needs to have tseries_depends, np? (e.g. np_datetime is there)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jbrockmendel** [comment]
> for tslibs.strptime, still needs to have tseries_depends, np? (e.g. np_datetime is there)

I thought we landed on not specifying recursive dependencies.  No?

File: setup.py (hunk 6 above)
Comment (jbrockmendel) [reply setup.py]: Well nat_strings is defined but not c-defined in nattype, so we could import it into parsing without needing to specify the dependency.

I'd be fine with that, or with having it defined one extra place.  parsing has no intra-pandas dependencies at the moment (neither import nor cimport) which makes it really low-stress.

**Comment by jreback** [comment]
> I thought we landed on not specifying recursive dependencies. No?

you mean transitive I suppose.

well this is a c-include and is needed for compilation. compile it in isolation and see if it works when you don't have these. anywhere we are implicity or explicity including np_datetime which is almost everywhere in tslibs these need to be specified, unless you can prove unambiguously that they are not needed.

**Comment by jreback** [comment]
I still also dont' understand why this is actually needed. what is broken if this is not included (the status quo)?

**Comment by jbrockmendel** [comment]
> you mean transitive I suppose.

Sure.  To be extra-clear: if A cimports from B and B cimports or "cdef extern"s from C, then A transitively/recursively depends on C, but does not directly depend on C.  Same page?

**Comment by jbrockmendel** [comment]
> anywhere we are implicity or explicity including np_datetime which is almost everywhere in tslibs these need to be specified, 

My understanding was that the policy is to specify only direct dependencies.  Otherwise, say, why  aren't hashtable's dependencies specified in all of the other extensions that cimport from hashtable?

**Comment by jbrockmendel** [comment]
> well this is a c-include and is needed for compilation. compile it in isolation and see if it works when you don't have these. 

Yah this gets back to the original thing of me not understanding why you wanted to add a bunch of junk to these specifications in the first place.  As long as direct dependencies are listed it compiles fine.  (wouldn't the CI here scoff otherwise?)

**Comment by jbrockmendel** [comment]
> unless you can prove unambiguously that they are not needed.

https://groups.google.com/forum/#!topic/cython-users/YdGpGqi6Qw4

**Comment by jbrockmendel** [comment]
> I still also dont' understand why this is actually needed. what is broken if this is not included (the status quo)?

1) The issue is that _too much_ is specified, and it is specified in _too many_ places, which is a recipe for confusion.

Explicitly setting the include key to either `[]` or `numpy_incls` where possible pins that down to the minimal viable set of dependencies.

**Comment by jbrockmendel** [comment]
The `numpy_incls` part of the PR is largely orthogonal to the discussion about transitive dependencies.  How about I revert the changes that affect transitive dependencies and we can address that elsewhere/elsewhen?

**Comment by jbrockmendel** [comment]
Just pushed a commit reverting the subset of changes on which there has been controversy.

Now this just sets the `include` key to either `[]` or `numpy_incls` where applicable, ignores specification (or possible over-specification) of transitive dependencies.

**Comment by jbrockmendel** [comment]
Clipboard.

**Comment by jreback** [comment]
> As long as direct dependencies are listed it compiles fine. (wouldn't the CI here scoff otherwise?)

well the CI won't complain, but building from a previous state, DOES complain. I don't want to have to constantly ``make clean``; the current building is actually nice, if a .pxd changes or a dep changes, then it only recompiles what is necessary.

So back to this PR. I don't see any effect (pos or neg), so what exactly is symptom that this is trying to fix?

**Comment by jbrockmendel** [comment]
> So back to this PR. I don't see any effect (pos or neg), so what exactly is symptom that this is trying to fix?

Specifying dependencies (or "includes") which the relevant extensions are not actually dependent on is an anti-pattern.  This PR fixes that for the subset of pyx files that have zero intra-pandas dependencies.

**Comment by jbrockmendel** [comment]
On green let’s either merge or kill, avoid clogging the CI queue.  Especially in light of reopening the depends issue, I think this is worthwhile.

**Comment by jbrockmendel** [comment]
Doing some googling on what affects compile/build times, one thing that came up frequently was unnecessary includes.  No idea how applicable it is here, but its plausible that trimming unnecessary includes and sources could cut build time.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
ok thanks. We do need to refactor ``setup.py`` to use ``cythonize`` that will reduce the code a lot, while preserving the dependencies (w/o us having to do anything).

**Comment by jreback** [comment]
so I am pretty sure this broke our wheel building for 2.7, see here: https://travis-ci.org/MacPython/pandas-wheels/jobs/304148523

this is not easy to try a fix, and we have to commit it and see (though you could try forking that repo if you want).

**Comment by jbrockmendel** [comment]
Huh.

> ImportError: No module named _build_utils.apple_accelerate

**Pull request merged**